### PR TITLE
docs - CTE available with INSERT, UPDATE, DELETE

### PR DESCRIPTION
**Pull request by mkiyama**
-updated GUC
-updated Admin Guide topic WITH Queries (Common Table Expressions)

updates to SELECT, INSERT, UPDATE DELETE references will be part of postgres 9.2 doc merge.

### CODE REVIEW

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 1]
```diff
@@ -3,14 +3,55 @@
...
-      thought of as defining temporary tables that exist just for the query. One use of this feature
-      is to break down complicated queries into simpler parts. This example query displays
-      per-product sales totals in only the top sales regions:</p>
-    <codeblock>WITH regional_sales AS (
+    <p>See <xref href="#topic_xyn_dgh_5gb" format="dita"/> for information about using
+        <codeph>SELECT</codeph> in a <codeph>WITH</codeph> clause.</p>
+    <p>See <xref href="#topic_zg3_bgh_5gb" format="dita"/>, for information about using
+        <codeph>INSERT</codeph>, <codeph>UPDATE</codeph>, or <codeph>DELETE</codeph> in a
+        <codeph>WITH</codeph> clause. </p>
+    <note>These are limitations for using a <codeph>WITH</codeph> clause.<p>For a
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: Make this a bullet list of limitations?

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 2]
```diff
@@ -3,14 +3,55 @@
...
+        not intended for use in a production environment. Experimental features are subject to
+        change without notice in future releases.</note></p>
+  </body>
+  <topic id="topic_xyn_dgh_5gb">
+    <title>SELECT in a WITH Clause</title>
+    <body>
+      <p>The subqueries, which are often referred to as Common Table Expressions or CTEs, can be
+        thought of as defining temporary tables that exist just for the query. These examples show
+        the <codeph>WITH</codeph> clause being used with a <codeph>SELECT</codeph> command. The
+        example <codeph>WITH</codeph> clause can used the same way with <codeph>INSERT</codeph>,
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: clause -> clauses

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 3]
```diff
@@ -3,14 +3,55 @@
...
+          <codeph>UPDATE</codeph>, or <codeph>DELETE</codeph>. In each case, it effectively provides
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: it -> the WITH clause

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 4]
```diff
@@ -3,14 +3,55 @@
...
+        temporary tables that can be referred to in the main command.</p>
+      <p>A <codeph>SELECT</codeph> command in the <codeph>WITH</codeph> clause is evaluated only
+        once per execution of the parent query, even if it is referred to more than once by the
+        parent query or sibling <codeph>WITH</codeph> clauses. Thus, expensive calculations that are
+        needed in multiple places can be placed within a <codeph>WITH</codeph> clause to avoid
+        redundant work. Another possible application is to prevent unwanted multiple evaluations of
+        functions with side-effects. However, the other side of this coin is that the optimizer is
+        less able to push restrictions from the parent query down into a WITH query than an ordinary
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: missing codeph for WITH here

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 5]
```diff
@@ -27,50 +68,49 @@ FROM orders
...
-      the <codeph>WITH</codeph> clause can refer to its own output. This is a simple example that
-      computes the sum the integers from 1 through 100:</p>
-    <codeblock>WITH RECURSIVE t(n) AS (
+      <p>The query could have been written without the <codeph>WITH</codeph> clause, but would have
+        required two levels of nested sub-SELECTs. It is easier to follow with the
+          <codeph>WITH</codeph> clause.</p>
+      <p>When the optional <codeph>RECURSIVE</codeph> keyword is enabled, the <codeph>WITH</codeph>
+        clause can accomplish things not otherwise possible in standard SQL. Using
+          <codeph>RECURSIVE</codeph>, a query in the <codeph>WITH</codeph> clause can refer to its
+        own output. This is a simple example that computes the sum the integers from 1 through
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: the sum the -> the sum of

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 6]
```diff
@@ -27,50 +68,49 @@ FROM orders
...
-        </ol></li>
-    </ol>
-    <note>Strictly speaking, the process is iteration not recursion, but <codeph>RECURSIVE</codeph>
-      is the terminology chosen by the SQL standards committee.</note>
-    <p>Recursive <codeph>WITH</codeph> queries are typically used to deal with hierarchical or
-      tree-structured data. An example is this query to find all the direct and indirect sub-parts
-      of a product, given only a table that shows immediate inclusions:</p>
-    <codeblock>WITH RECURSIVE included_parts(sub_part, part, quantity) AS (
+      <p>The general form of a recursive <codeph>WITH</codeph> clause is a <i>non-recursive
+        term</i>, followed a <codeph>UNION</codeph> (or <codeph>UNION ALL</codeph>), and then a
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: followed a -> followed by a

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 7]
```diff
@@ -81,19 +121,20 @@ SELECT sub_part, SUM(quantity) as total_quantity
...
-      ensure that the recursive part of the query eventually returns no tuples by discarding rows
-      that duplicate previous output rows. However, often a cycle does not involve output rows that
-      are complete duplicates: it might sufficient to check just one or a few fields to see if the
-      same point has been reached before. The standard method for handling such situations is to
-      compute an array of the visited values. For example, consider the following query that
-      searches a table graph using a link field:</p>
-    <codeblock>WITH RECURSIVE search_graph(id, link, data, depth) AS (
+      <p>When working with recursive <codeph>WITH</codeph> queries, you must ensure that the
+        recursive part of the query eventually returns no tuples, or else the query loops
+        indefinitely. In the example that computes the sum the integers, the working table contains
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: the sum the -> the sum of

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 8]
```diff
@@ -4379,15 +4379,17 @@
     <title>gp_recursive_cte_prototype</title>
     <body>
       <p>Controls the availability of the experimental keyword <codeph>RECURSIVE</codeph> in the
-          <codeph>WITH</codeph> clause of a <codeph>SELECT[ INTO]</codeph> command. The keyword
-        allows a subquery in the <codeph>WITH</codeph> clause of a <codeph>SELECT[ INTO]</codeph>
-        command to reference itself. The default value is <codeph>false</codeph>, the keyword is not
-        allowed in the <codeph>WITH</codeph> clause a <codeph>SELECT[ INTO]</codeph> command.</p>
+          <codeph>WITH</codeph> clause of a <codeph>SELECT[ INTO]</codeph> command, or a
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: SELECT[ INTO] -> SELECT [INTO]

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 9]
```diff
@@ -3,14 +3,59 @@
...
+        not intended for use in a production environment. Experimental features are subject to
+        change without notice in future releases.</note></p>
+  </body>
+  <topic id="topic_xyn_dgh_5gb">
+    <title>SELECT in a WITH Clause</title>
+    <body>
+      <p>The subqueries, which are often referred to as Common Table Expressions or CTEs, can be
+        thought of as defining temporary tables that exist just for the query. These examples show
+        the <codeph>WITH</codeph> clause being used with a <codeph>SELECT</codeph> command. The
+        example <codeph>WITH</codeph> clauses can used the same way with <codeph>INSERT</codeph>,
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: can *be* used?

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 10]
```diff
@@ -81,19 +125,20 @@ SELECT sub_part, SUM(quantity) as total_quantity
...
+      <p>When working with recursive <codeph>WITH</codeph> queries, you must ensure that the
+        recursive part of the query eventually returns no tuples, or else the query loops
+        indefinitely. In the example that computes the sum of integers, the working table contains a
+        single row in each step, and it takes on the values from 1 through 100 in successive steps.
+        In the 100th step, there is no output because of the <codeph>WHERE</codeph> clause, and the
+        query terminates.</p>
+      <p>For some queries, using <codeph>UNION</codeph> instead of <codeph>UNION ALL</codeph> can
+        ensure that the recursive part of the query eventually returns no tuples by discarding rows
+        that duplicate previous output rows. However, often a cycle does not involve output rows
+        that are complete duplicates: it might sufficient to check just one or a few fields to see
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: might *be* sufficient?

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 11]
```diff
@@ -103,12 +148,13 @@ GROUP BY sub_part
...
-        <codeph>depth</codeph> output, changing <codeph>UNION ALL</codeph> to <codeph>UNION</codeph>
-      does not eliminate the looping. Instead the query needs to recognize whether it has reached
-      the same row again while following a particular path of links. This modified query adds two
-      columns <codeph>path</codeph> and <codeph>cycle</codeph> to the loop-prone query:</p>
-    <codeblock>WITH RECURSIVE search_graph(id, link, data, depth, path, cycle) AS (
+      <p>This query loops if the link relationships contain cycles. Because the query requires a
+          <codeph>depth</codeph> output, changing <codeph>UNION ALL</codeph> to
+          <codeph>UNION</codeph> does not eliminate the looping. Instead the query needs to
+        recognize whether it has reached the same row again while following a particular path of
+        links. This modified query adds two columns <codeph>path</codeph> and <codeph>cycle</codeph>
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: two columns *,* path and cycle *,* ?

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 12]
```diff
@@ -141,25 +187,95 @@ SELECT * FROM search_graph;
...
-    <note type="tip">The recursive query evaluation algorithm produces its output in breadth-first
-      search order. You can display the results in depth-first search order by making the outer
-      query <codeph>ORDER BY</codeph> a path column constructed in this way.</note>
-    <p>A helpful technique for testing a query when you are not certain if it might loop
-      indefinitely is to place a <codeph>LIMIT</codeph> in the parent query. For example, this query
-      would loop forever without the <codeph>LIMIT</codeph> clause:</p>
-    <codeblock>WITH RECURSIVE t(n) AS (
+      <note type="tip">Omit the <codeph>ROW()</codeph> syntax in the case where only one field needs
+        to be checked to recognize a cycle. This uses a simple array rather than a composite-type
+        array to be used, gaining efficiency.</note>
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: remove "to be used"?

File: gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml [hunk 13]
```diff
@@ -141,25 +187,95 @@ SELECT * FROM search_graph;
...
+      <p>Data-modifying statements in a <codeph>WITH</codeph> clause must have
+          <codeph>RETURNING</codeph> clauses, as shown in the previous example. It is the output of
+        the <codeph>RETURNING</codeph> clause, not the target table of the data-modifying statement,
+        that forms the temporary table that can be referred to by the rest of the query. If a
+        data-modifying statement in a <codeph>WITH</codeph> lacks a <codeph>RETURNING</codeph>
+        clause, an error is returned. </p>
+      <p>If the optional <codeph>RECURSIVE</codeph> keyword is enabled, recursive self-references in
+        data-modifying statements are not allowed. In some cases it is possible to work around this
+        limitation by referring to the output of a recursive <codeph>WITH</codeph>. For example,
+        this query would remove all direct and indirect subparts of a product.</p>
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: codeph-ify sub_parts?

Comment (mkiyama) [reply gpdb-doc/dita/admin_guide/query/topics/CTE-query.xml]: kept as is for the term subpart, not the database column.

**Pull request merged**